File: 8/9.js
// 시간 복잡도: O(n^m)
// 공간 복잡도: O(m)
function main(n, arr, m) {
  let result = 500;

  function dfs(level, sum) {
    if (sum > m) return;
    if (level > n) return;
    if (sum === m) {
      result = Math.min(result, level);
      return;
    }

    for (let i = 0; i < n; i++) {
      dfs(level + 1, sum + arr[i]);
    }
  }

  dfs(0, 0);

  return result;
}

console.log(main(3, [1, 2, 5], 15));

File: 8/9.py
def main(n, arr, m):
    result = 500

    def dfs(level, sum):
        nonlocal result

        if sum > m:
            return
        if level > n:
            return
        if sum == m:
            result = min(result, level)
            return

        for i in range(n):
            dfs(level + 1, sum + arr[i])

    dfs(0, 0)
    return result


print(main(3, [1, 2, 5], 15))
